refactor!: merging event emission functions

benesjan opened the pull request:
## Summary
This PR consolidates the event emission API by merging `emit_event_in_private_log` and `emit_event_as_offchain_message` into a single `emit_event_in_private` function with configurable delivery modes.

## Key Changes
- **Unified API**: Replaced `PrivateLogContentConstraintsEnum` with `MessageDeliveryEnum` that provides clear delivery mode options
- **Two Primary Modes**:
  - `CONSTRAINED_ONCHAIN`: Constrained encryption + tagging + private log stream (for smart contract decisions)
  - `UNCONSTRAINED_OFFCHAIN`: Unconstrained encryption + offchain delivery (for trusted/incentivized scenarios)
- **Enhanced Documentation**: Comprehensive documentation explaining the three dimensions of message delivery (encryption constraints, delivery mechanism, tagging constraints) and their use cases
- **Backward Compatibility**: Maintained `UNCONSTRAINED_ONCHAIN` mode for existing code (marked for future removal)

## Benefits
- **Clearer Intent**: Developers can now choose delivery modes based on their specific security and trust requirements
- **Simplified API**: Single function handles all private event emission scenarios
- **Improved Security Model**: Explicit guidance on when to use each delivery mode

## Breaking Changes
- `emit_event_in_private_log` → `emit_event_in_private`
- `emit_event_as_offchain_message` → `emit_event_in_private` with `UNCONSTRAINED_OFFCHAIN` mode
- `PrivateLogContent` → `MessageDelivery`

All contract examples and tests have been updated to use the new API.

benesjan commented:
* **#16126** <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16126?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/AztecProtocol/aztec-packages/16126?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -14,50 +14,134 @@ use dep::protocol_types::{
+///
+/// The following section outlines the recommended delivery variants and when to use each one.
+/// Each variant represents a different tradeoff between security guarantees, gas costs, and trust assumptions.
+/// - We are currently still exploring the design space but it seems that there are only 2 meaningful combinations:
```
I wonder if the reviewer agrees with this. I feel like we are converging only on 2 meaningful configurations here.

I can imagine that a dev might want to use unconstrained encryption, unconstrained tagging and private log stream if he wants it cheap but is lazy to implement the offchain message delivery in the app. However I think in that case we should spin up some centralized service such that this is simple to do and headache-free for devs. If we do this then there really doesn't seem to be any reason for existence of other combinations.

benesjan commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -71,48 +157,23 @@ where
+    }
 }
 
+// TODO: rename to emit_event_in_public
```
The naming is now inconsistent. Would the reviewer agree with the rename? If yes, will tackle in a followup PR.

nventuro replied:
It's just the `_log` suffix no? Sure, I don't think it's very important. We may want to change these to be methods soon anyway.

benesjan replied:
> Sure, I don't think it's very important.

I thought you are a "cute code" kind of person 😔

> We may want to change these to be methods soon anyway.

WDYM by this? Methods of context?

nventuro replied:
Methods of the event type. So you'd do something like `event.emit` instead of `emit_event(event)`

benesjan commented on yarn-project/end-to-end/src/e2e_offchain_effect.test.ts:
```diff
@@ -72,7 +72,7 @@ describe('e2e_offchain_effect', () => {
     const offchainEffect = offchainEffects[0];
 
-    // The data contains the cyphertext, an identifier and the recipient
+    // The data contains the ciphertext, an identifier and the recipient
```
@Thunkar ser, pls have mercy with my autism

https://marketplace.cursorapi.com/items/?itemName=streetsidesoftware.code-spell-checker-cspell-bundled-dictionaries

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -14,50 +14,136 @@ use dep::protocol_types::{
 
-pub struct PrivateLogContentConstraintsEnum {
-    /// The contents of the log are entirely unconstrained, and could have any values.
+/// Specifies the configuration parameters for message delivery across three key dimensions:
```
holy comments batman @iAmMichaelConnor this you?

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -14,50 +14,136 @@ use dep::protocol_types::{
-    CONSTRAINED_ENCRYPTION: 2,
-    // TODO: add constrained tagging and constrained handshaking
+pub global MessageDelivery: MessageDeliveryEnum = MessageDeliveryEnum {
+    CONSTRAINED_ONCHAIN: 1,
+    UNCONSTRAINED_OFFCHAIN: 2,
+    UNCONSTRAINED_ONCHAIN: 3,
```
This ordering is odd, I'd expect for this to be ordered with either increasing or decreasing guarantees, but this is neither. 1 is most guarantees, 2 is least guarantees, 3 is in the middle.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/16126/commits/095d1c561102b6451779d3adc83d14e8f2c85f89

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -14,50 +14,136 @@ use dep::protocol_types::{
+    assert_constant(delivery_mode);
+
+    // The following maps out the 3 dimensions across which we configure message delivery.
+    let unconstrained_encryption = delivery_mode != MessageDelivery.CONSTRAINED_ONCHAIN;
```
I think this would read better with equality and ORs, as you'd see more clearly what each variant does. E.g.

```noir
let unconstrained_encryption = delivery_mode == MessageDelivery.UNCONSTRAINED_OFFCHAIN | UNCONSTRAINED_ONCHAIN;
```

it also avoid the double negation in the `!=` and `unconstrained_enc`. It'd likely be even better if we had 


```noir
let constrained_encryption = delivery_mode == MessageDelivery.CONSTRAINED_ONCHAIN;
```

nventuro replied:
Other than that these booleans are nice.

benesjan replied:
Changing that to `constrained_encryption` indeed [looks nicer](https://github.com/AztecProtocol/aztec-packages/pull/16126/commits/6b9306c163928c730d6cb879f639be0c10c6146c). Thanks

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -71,48 +157,23 @@ where
-    context.push_nullifier(event_commitment);
+        // Safety: Offchain messages are by definition unconstrained. They are emitted via the `emit_offchain_effect`
+        // oracle which we don't use for anything besides its side effects, therefore this is safe to call.
+        unsafe { emit_offchain_message(log_content, recipient) };
```
`emit_offchain_message` should likely not be an unconstrained function. Also, why is this called `log_content`? It's the message ciphertext, there is no log.

benesjan replied:
>emit_offchain_message should likely not be an unconstrained function. 

will tackle in a followup PR

> why is this called log_content

https://github.com/AztecProtocol/aztec-packages/pull/16126/commits/20a260570eb5c246d80534c8bd962b4f3b9d288c

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -71,48 +157,23 @@ where
+        // oracle which we don't use for anything besides its side effects, therefore this is safe to call.
+        unsafe { emit_offchain_message(log_content, recipient) };
+    } else {
+        // When delivering through the private log stream, tagging is used so we must prefix the ciphertext with a tag.
```
log stream?

nventuro replied:
If we're calling this option `CONSTRAINED_ONCHAIN` then there should be a massive TODO here pointing out that delivery is not actually being constrained because `prefix_with_tag` does not constrain the tag. In fact I'd delete prefix with tag so that we can clearly see that the prefixing _is_ constrained, but the tag is not. And have the `Safety:` comment explain that this is not in fact safe. Alternatively we rename this to something else until we do have constr delivery.

benesjan replied:
Since we will need to tackle constrained tagging before mainnet think it makes sense to aim for a final API here and just document this. Added there a [comment](https://github.com/AztecProtocol/aztec-packages/pull/16126/commits/60c6a324f63a01adf01e3c225b493096f83d0ddd) documenting my lies and also added a security label to [the issue](https://github.com/AztecProtocol/aztec-packages/issues/14565).

nventuro commented on noir-projects/noir-contracts/contracts/app/simple_token_contract/src/main.nr:
```diff
@@ -153,11 +153,11 @@ pub contract SimpleToken {
             &mut context,
             to,
-            PrivateLogContent.NO_CONSTRAINTS,
+            MessageDelivery.UNCONSTRAINED_ONCHAIN,
```
This does read much better, yeah. We may want to iterate on the names of the variants.

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -14,50 +14,136 @@ use dep::protocol_types::{
+/// +--------------------------------------------------------------------------------------------------------+
+///
+///
+/// # Message authenticity
+/// Note that we always commit to the plaintext and the recipient is therefore always able to verify
+/// the authenticity of the message no matter the delivery mechanism.
```
This is false - there is no plaintext commitment, and with unconstrained encryption the plaintext is arbitrary. we do have commitments to events and notes, but those are very different things

benesjan replied:
So your argument is that the `MessageDeliveryEnum` is a generic thing independent of events and notes and hence the statement is incorrectly placed here?

If yes that makes sense to me. Had notes and events in mind when I was writing this. Nuked the paragraph in https://github.com/AztecProtocol/aztec-packages/pull/16126/commits/a22f523dc50cae4c9921bbf8055f1ef7e5bd98cd

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -14,50 +14,136 @@ use dep::protocol_types::{
+/// +--------------------------------------------------------------------------------------------------------+
+///
+///
+/// Other combinations like unconstrained encryption, unconstrained tagging, private log stream don't seem
+/// to be meaningful because they are costlier than combination 2 above while providing no meaningful
+/// additional guarantees.
```
Unconstrained onchain delivery is very much relevant. it means you don't need any third party solution to handle the delivery, you don't need servers, an out of band channel, backups, nothing. I can just send you tokens from my wallet and it'll work with zero risk. It is very valid.

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/commit/6c585466d2fe8e20e9b9df120daea939f44b4e83

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -14,50 +14,136 @@ use dep::protocol_types::{
-    pub NO_CONSTRAINTS: u8,
-    /// The contents of the log and its encryption are constrained. The tag (and therefore whether the recipient is
-    /// actually able to find the message) is not.
+    /// This gives us an on-chain guarantee that a recipient is able to discover and receive a message. See
+    /// combination 1 above for details.
+    pub CONSTRAINED_ONCHAIN: u8,
```
This is not incorrect, but it's a lie - we do not support this at all. We don't have a way to do constrained tagging. It is being designed and still under discussion. Sure, that's what we want to convey I guess. It's just not true right now

iAmMichaelConnor replied:
Oooh an opportunity to shill one of my hackmds! Jan, check this out: https://hackmd.io/MY7QqIaxTg-a2CMu_Znu4Q
You'll love it. Better than Harry Potter.

benesjan replied:
> This is not incorrect, but it's a lie - we do not support this at all. 

Clarified this in https://github.com/AztecProtocol/aztec-packages/pull/16126/commits/6c585466d2fe8e20e9b9df120daea939f44b4e83


@iAmMichaelConnor Very cute hackmd Mike. You definitely deserve a 🍪

In the "Why Constrained Delivery?" section you wanted to come up with more examples. Hackmd is buggy for me and I just didn't manage to comment on that. Another example could be what I just wrote in this PR:

"An escrow contract facilitating a private NFT sale that needs to verify payment before releasing the NFT to the buyer."

nventuro commented on noir-projects/aztec-nr/aztec/src/event/event_interface.nr:
```diff
@@ -14,50 +14,136 @@ use dep::protocol_types::{
 
-pub struct PrivateLogContentConstraintsEnum {
-    /// The contents of the log are entirely unconstrained, and could have any values.
+/// Specifies the configuration parameters for message delivery across three key dimensions:
```
Some extra thoughts re. this text. There are some inaccuracies, but other than that it is not incorrect. I don't think however that it is very educational - even the ordering of concepts is odd. For example, I'd first mention that delivery can be done on chain or out of band, and that if done out of band no constraints can exist on the message because it is never stored anywhere anyway. Then we can talk about encryption for the on-chain case, which is the only place where it is relevant. But doing encryption and then delivery just seems odd.

Then there's the second section at the end, which just repeats the three decisions that are explained here, except onchain unconstrained is mistakenly discarded as useless. Both can just be merged.

benesjan replied:
Good points. Addressed in https://github.com/AztecProtocol/aztec-packages/pull/16126/commits/6c585466d2fe8e20e9b9df120daea939f44b4e83

nventuro reviewed (approved):
Thanks a lot for the responses to the review comments!

Commit:
refactor!: merging event emission functions

## Summary
This PR consolidates the event emission API by merging `emit_event_in_private_log` and `emit_event_as_offchain_message` into a single `emit_event_in_private` function with configurable delivery modes.

## Key Changes
- **Unified API**: Replaced `PrivateLogContentConstraintsEnum` with `MessageDeliveryEnum` that provides clear delivery mode options
- **Two Primary Modes**:
  - `CONSTRAINED_ONCHAIN`: Constrained encryption + tagging + private log stream (for smart contract decisions)
  - `UNCONSTRAINED_OFFCHAIN`: Unconstrained encryption + offchain delivery (for trusted/incentivized scenarios)
- **Enhanced Documentation**: Comprehensive documentation explaining the three dimensions of message delivery (encryption constraints, delivery mechanism, tagging constraints) and their use cases
- **Backward Compatibility**: Maintained `UNCONSTRAINED_ONCHAIN` mode for existing code (marked for future removal)

## Benefits
- **Clearer Intent**: Developers can now choose delivery modes based on their specific security and trust requirements
- **Simplified API**: Single function handles all private event emission scenarios
- **Improved Security Model**: Explicit guidance on when to use each delivery mode

## Breaking Changes
- `emit_event_in_private_log` → `emit_event_in_private`
- `emit_event_as_offchain_message` → `emit_event_in_private` with `UNCONSTRAINED_OFFCHAIN` mode
- `PrivateLogContent` → `MessageDelivery`

All contract examples and tests have been updated to use the new API.
